File: samples/go/workflow_overview.go
// Simplified Make connection workflow in Go.
//
// This single file mirrors the Node sample so teams can understand the control flow,
// API payloads, and blueprint rewriting without digging through a larger codebase.
//
// Usage:
//   go run workflow_overview.go "Hubspot - FBcapi.json" "Optional Account Name"
package main

import (
    "bufio"
    "bytes"
    "encoding/json"
    "errors"
    "fmt"
    "io"
    "net/http"
    "net/url"
    "os"
    "os/exec"
    "path/filepath"
    "runtime"
    "strings"
)

var rootDir = func() string {
    dir, _ := filepath.Abs(filepath.Join(filepath.Dir(os.Args[0]), "..", ".."))
    return dir
}()

func main() {
    if len(os.Args) < 2 {
        fmt.Println("Usage: go run workflow_overview.go <blueprint.json> [account-name]")
        os.Exit(1)
    }

    blueprintName := os.Args[1]
    accountOverride := ""
    if len(os.Args) > 2 {
        accountOverride = os.Args[2]
    }

    env, err := loadEnv(filepath.Join(rootDir, ".env"))
    if err != nil {
        panic(err)
    }

    settings := Settings{
        InstanceURL:  firstNonEmpty(env["INSTANCE_URL"], os.Getenv("INSTANCE_URL")),
        AuthToken:    firstNonEmpty(env["AUTH_TOKEN"], os.Getenv("AUTH_TOKEN")),
        TeamID:       firstNonEmpty(env["TEAM_ID"], os.Getenv("TEAM_ID")),
        BlueprintDir: filepath.Join(rootDir, "blueprints"),
        UpdatedDir:   filepath.Join(rootDir, "samples", "go"),
    }

    if err := ensureSettings(settings); err != nil {
        panic(err)
    }

    blueprint, blueprintPath, err := loadBlueprint(settings, blueprintName)
    if err != nil {
        panic(err)
    }

    appName, modules := analyseBlueprint(blueprint)
    appDefinition, err := fetchJSON(settings, fmt.Sprintf("/api/v2/imt/apps/%s", appName), "GET", nil)
    if err != nil {
        panic(err)
    }
    scopes := collectScopes(appDefinition, modules)
    formSchema, err := fetchJSON(settings, fmt.Sprintf("/api/v2/imt-forms/connections/create?type=%s&teamId=%s", appName, settings.TeamID), "GET", nil)
    if err != nil {
        panic(err)
    }
    payload := buildPayload(formSchema, appName, scopes, accountOverride)

    createResp, err := fetchJSON(settings, fmt.Sprintf("/api/v2/connections?teamId=%s&inspector=0", settings.TeamID), "POST", payload)
    if err != nil {
        panic(err)
    }
    connection, ok := createResp["connection"].(map[string]interface{})
    if !ok {
        connection = createResp
    }
    if _, ok := connection["id"]; !ok {
        panic("connection response missing id")
    }

    consentURL, err := fetchConsentURL(settings, connection["id"])
    if err != nil {
        panic(err)
    }

    rewritten, updated := rewriteBlueprint(blueprint, connection["id"])
    path, err := saveBlueprint(settings, blueprintPath, rewritten)
    if err != nil {
        panic(err)
    }

    fmt.Println("Connection created:", connection["id"])
    fmt.Println("Consent URL:", consentURL)
    fmt.Printf("Updated blueprint (%d modules): %s\n", updated, path)

    openBrowser(consentURL)
    fmt.Print("Complete consent, then press Enter to run the connection test...")
    bufio.NewReader(os.Stdin).ReadBytes('\n')

    testResp, err := fetchJSON(settings, fmt.Sprintf("/api/v2/connections/%v/test", connection["id"]), "POST", nil)
    if err != nil {
        panic(err)
    }
    pretty, _ := json.MarshalIndent(testResp, "", "  ")
    fmt.Println(string(pretty))
}

// Settings keeps configuration together so we do not pass dozens of arguments around.
type Settings struct {
    InstanceURL  string
    AuthToken    string
    TeamID       string
    BlueprintDir string
    UpdatedDir   string
}

// loadEnv reads a minimal .env file (key=value) without external dependencies.
func loadEnv(path string) (map[string]string, error) {
    env := make(map[string]string)
    file, err := os.Open(path)
    if err != nil {
        if errors.Is(err, os.ErrNotExist) {
            return env, nil
        }
        return nil, err
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        line := strings.TrimSpace(scanner.Text())
        if line == "" || strings.HasPrefix(line, "#") || !strings.Contains(line, "=") {
            continue
        }
        parts := strings.SplitN(line, "=", 2)
        env[strings.TrimSpace(parts[0])] = strings.Trim(strings.TrimSpace(parts[1]), "\"")
    }
    return env, scanner.Err()
}

// ensureSettings validates that the inputs required by every Make API call exist.
func ensureSettings(settings Settings) error {
    missing := make([]string, 0)
    if settings.InstanceURL == "" {
        missing = append(missing, "INSTANCE_URL")
    }
    if settings.AuthToken == "" {
        missing = append(missing, "AUTH_TOKEN")
    }
    if settings.TeamID == "" {
        missing = append(missing, "TEAM_ID")
    }
    if len(missing) > 0 {
        return fmt.Errorf("missing required env vars: %s", strings.Join(missing, ", "))
    }
    return nil
}

// fetchJSON performs an HTTP request against the Make API and decodes JSON responses.
func fetchJSON(settings Settings, path string, method string, body interface{}) (map[string]interface{}, error) {
    url := strings.TrimSuffix(settings.InstanceURL, "/") + path
    var reqBody io.Reader
    if body != nil {
        raw, err := json.Marshal(body)
        if err != nil {
            return nil, err
        }
        reqBody = bytes.NewReader(raw)
    }
    req, err := http.NewRequest(method, url, reqBody)
    if err != nil {
        return nil, err
    }
    req.Header.Set("Content-Type", "application/json")
    req.Header.Set("Authorization", "Token "+settings.AuthToken)

    resp, err := http.DefaultClient.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode >= 300 {
        bodyBytes, _ := io.ReadAll(resp.Body)
        return nil, fmt.Errorf("request failed (%d): %s", resp.StatusCode, string(bodyBytes))
    }

    var data map[string]interface{}
    decoder := json.NewDecoder(resp.Body)
    if err := decoder.Decode(&data); err != nil {
        return nil, err
    }
    return data, nil
}

// fetchConsentURL requests the browser redirect URL without following the redirect.
func fetchConsentURL(settings Settings, connectionID interface{}) (string, error) {
    path := fmt.Sprintf("/api/v2/oauth/auth/%v", connectionID)
    url := strings.TrimSuffix(settings.InstanceURL, "/") + path
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return "", err
    }
    req.Header.Set("Authorization", "Token "+settings.AuthToken)

    client := &http.Client{CheckRedirect: func(req *http.Request, via []*http.Request) error {
        return http.ErrUseLastResponse
    }}
    resp, err := client.Do(req)
    if err != nil {
        return "", err
    }
    defer resp.Body.Close()

    if resp.StatusCode < 300 || resp.StatusCode >= 400 {
        return "", fmt.Errorf("unexpected status %d when requesting consent", resp.StatusCode)
    }
    location := resp.Header.Get("Location")
    if location == "" {
        return "", errors.New("missing Location header")
    }
    return location, nil
}

// loadBlueprint accepts either an absolute path or a filename inside the blueprint directory.
func loadBlueprint(settings Settings, name string) (map[string]interface{}, string, error) {
    candidates := []string{
        name,
        filepath.Join(settings.BlueprintDir, name),
    }
    for _, candidate := range candidates {
        if data, err := os.ReadFile(candidate); err == nil {
            var blueprint map[string]interface{}
            if err := json.Unmarshal(data, &blueprint); err != nil {
                return nil, "", err
            }
            return blueprint, candidate, nil
        }
    }
    return nil, "", fmt.Errorf("blueprint not found: %s", name)
}

// rewriteBlueprint clones the blueprint and swaps __IMTCONN__ for every non-Facebook module.
func rewriteBlueprint(blueprint map[string]interface{}, connectionID interface{}) (map[string]interface{}, int) {
    raw, _ := json.Marshal(blueprint)
    clone := make(map[string]interface{})
    json.Unmarshal(raw, &clone)
    updated := 0

    var walk func(interface{})
    walk = func(node interface{}) {
        switch v := node.(type) {
        case []interface{}:
            for _, item := range v {
                walk(item)
            }
        case map[string]interface{}:
            if module, ok := v["module"].(string); ok && !strings.HasPrefix(module, "facebook-conversion-leads") {
                if params, ok := v["parameters"].(map[string]interface{}); ok {
                    if _, exists := params["__IMTCONN__"]; exists {
                        params["__IMTCONN__"] = connectionID
                        updated++
                    }
                }
            }
            for _, value := range v {
                walk(value)
            }
        }
    }

    walk(clone["flow"])
    return clone, updated
}

// saveBlueprint writes the rewritten JSON next to the sample file to keep everything portable.
func saveBlueprint(settings Settings, original string, blueprint map[string]interface{}) (string, error) {
    if err := os.MkdirAll(settings.UpdatedDir, 0o755); err != nil {
        return "", err
    }
    base := strings.TrimSuffix(filepath.Base(original), filepath.Ext(original))
    path := filepath.Join(settings.UpdatedDir, base+"-updated.json")
    data, err := json.MarshalIndent(blueprint, "", "  ")
    if err != nil {
        return "", err
    }
    if err := os.WriteFile(path, append(data, '\n'), 0o644); err != nil {
        return "", err
    }
    return path, nil
}

// openBrowser triggers the platform specific command to launch the consent URL.
func openBrowser(rawURL string) {
    if _, err := url.ParseRequestURI(rawURL); err != nil {
        fmt.Println("Invalid consent URL:", rawURL)
        return
    }
    var cmd *exec.Cmd
    switch runtime.GOOS {
    case "darwin":
        cmd = exec.Command("open", rawURL)
    case "windows":
        cmd = exec.Command("cmd", "/c", "start", rawURL)
    default:
        cmd = exec.Command("xdg-open", rawURL)
    }
    _ = cmd.Start()
}

// analyseBlueprint extracts the target app and module names from the blueprint tree.
func analyseBlueprint(blueprint map[string]interface{}) (string, []string) {
    modules := make([][2]string, 0)
    var walk func(interface{})
    walk = func(node interface{}) {
        switch v := node.(type) {
        case []interface{}:
            for _, item := range v {
                walk(item)
            }
        case map[string]interface{}:
            if module, ok := v["module"].(string); ok {
                parts := strings.SplitN(module, ":", 2)
                if len(parts) == 2 {
                    modules = append(modules, [2]string{parts[0], parts[1]})
                }
            }
            for _, value := range v {
                walk(value)
            }
        }
    }
    walk(blueprint["flow"])

    for _, pair := range modules {
        if pair[0] != "facebook-conversion-leads" {
            selectedApp := pair[0]
            moduleNames := make([]string, 0)
            for _, candidate := range modules {
                if candidate[0] == selectedApp {
                    moduleNames = append(moduleNames, candidate[1])
                }
            }
            return selectedApp, moduleNames
        }
    }
    panic("no eligible modules found")
}

// collectScopes maps blueprint modules to scopes listed in the Make app definition.
func collectScopes(appDefinition map[string]interface{}, moduleNames []string) []string {
    lookup := make(map[string][]string)
    app := appDefinition["app"].(map[string]interface{})
    for _, bucket := range []string{"actions", "searches", "triggers"} {
        items, _ := app[bucket].([]interface{})
        for _, raw := range items {
            item := raw.(map[string]interface{})
            name, _ := item["name"].(string)
            if name == "" {
                continue
            }
            if scopes, ok := item["scopes"].([]interface{}); ok {
                lookup[strings.ToLower(name)] = toStringSlice(scopes)
            }
        }
    }
    combined := make(map[string]struct{})
    for _, name := range moduleNames {
        for _, scope := range lookup[strings.ToLower(name)] {
            combined[scope] = struct{}{}
        }
    }
    result := make([]string, 0, len(combined))
    for scope := range combined {
        result = append(result, scope)
    }
    return result
}

// buildPayload merges default form values with our account override and scoped permissions.
func buildPayload(schema map[string]interface{}, appName string, scopes []string, override string) map[string]interface{} {
    payload := make(map[string]interface{})
    fields := make([]map[string]interface{}, 0)

    var walk func(interface{})
    walk = func(node interface{}) {
        switch v := node.(type) {
        case []interface{}:
            for _, item := range v {
                walk(item)
            }
        case map[string]interface{}:
            key := ""
            if k, ok := v["key"].(string); ok && k != "" {
                key = k
            } else if name, ok := v["name"].(string); ok && name != "" {
                key = name
            }

            resolvedType := ""
            if t, ok := v["type"].(string); ok && t != "" {
                resolvedType = t
            } else if templ, ok := v["templateOptions"].(map[string]interface{}); ok {
                if t, ok := templ["type"].(string); ok {
                    resolvedType = t
                }
            }

            if key != "" {
                lowerType := strings.ToLower(resolvedType)
                if lowerType != "button" && lowerType != "content" && lowerType != "htmlelement" {
                    field := map[string]interface{}{
                        "key":              key,
                        "type":             resolvedType,
                        "data":             v["data"],
                        "templateOptions":  v["templateOptions"],
                        "defaultValue":     nil,
                    }
                    if defVal, ok := v["defaultValue"]; ok {
                        field["defaultValue"] = defVal
                    } else if templ, ok := v["templateOptions"].(map[string]interface{}); ok {
                        if dv, ok := templ["defaultValue"]; ok {
                            field["defaultValue"] = dv
                        } else if dv, ok := templ["default"]; ok {
                            field["defaultValue"] = dv
                        }
                    }
                    if field["defaultValue"] == nil {
                        if dv, ok := v["default"]; ok {
                            field["defaultValue"] = dv
                        }
                    }
                    fields = append(fields, field)
                }
            }

            for prop, value := range v {
                if prop == "components" {
                    continue
                }
                if prop == "options" {
                    if optMap, ok := value.(map[string]interface{}); ok {
                        if store, ok := optMap["store"].([]interface{}); ok && len(store) > 0 {
                            var selected map[string]interface{}
                            for _, candidate := range store {
                                option, ok := candidate.(map[string]interface{})
                                if !ok {
                                    continue
                                }
                                if def, ok := option["default"].(bool); ok && def {
                                    selected = option
                                    break
                                }
                                if sel, ok := option["selected"].(bool); ok && sel {
                                    selected = option
                                    break
                                }
                                if selected == nil {
                                    selected = option
                                }
                            }
                            if selected != nil {
                                walk(selected)
                            }
                            continue
                        }
                    }
                }
                if prop == "nested" {
                    if arr, ok := value.([]interface{}); ok {
                        for _, item := range arr {
                            walk(item)
                        }
                        continue
                    }
                }
                walk(value)
            }
        }
    }
    walk(schema)

    for _, field := range fields {
        key := field["key"].(string)
        if val, ok := field["defaultValue"]; ok && val != nil {
            payload[key] = val
        } else {
            payload[key] = fallbackForField(field)
        }
    }

    if name, ok := payload["accountName"].(string); ok {
        if strings.Contains(name, "{{") && strings.Contains(name, "}}") {
            payload["accountName"] = ""
        }
    }

    if payload["accountType"] == nil || payload["accountType"] == "" {
        payload["accountType"] = appName
    }
    if override != "" {
        payload["accountName"] = override
    } else if payload["accountName"] == nil || payload["accountName"] == "" {
        payload["accountName"] = appName + " connection"
    }
    payload["customScopes"] = scopes
    if val, ok := payload["property"]; ok && val == "" {
        panic("the connection form expects a property value")
    }
    return payload
}

// toStringSlice converts interface{} lists coming from JSON into []string for convenience.
func toStringSlice(items []interface{}) []string {
    result := make([]string, 0, len(items))
    for _, item := range items {
        if s, ok := item.(string); ok {
            result = append(result, s)
        }
    }
    return result
}

// firstNonEmpty returns the first non-empty string in the provided list.
func firstNonEmpty(values ...string) string {
    for _, v := range values {
        if strings.TrimSpace(v) != "" {
            return v
        }
    }
    return ""
}

func fallbackForField(field map[string]interface{}) interface{} {
    typeStr, _ := field["type"].(string)
    template, _ := field["templateOptions"].(map[string]interface{})
    if typeStr != "" && strings.EqualFold(typeStr, "boolean") {
        return false
    }
    if template != nil {
        if templType, _ := template["type"].(string); templType != "" && strings.EqualFold(templType, "boolean") {
            return false
        }
    }
    if value, ok := optionFallback(field, template); ok {
        return value
    }
    return ""
}

func optionFallback(field map[string]interface{}, template map[string]interface{}) (interface{}, bool) {
    sources := []map[string]interface{}{}
    if data, _ := field["data"].(map[string]interface{}); data != nil {
        sources = append(sources, data)
    }
    if template != nil {
        sources = append(sources, template)
    }
    if len(sources) == 0 {
        return nil, false
    }

    for _, source := range sources {
        if val, exists := source["defaultValue"]; exists {
            return val, true
        }
    }

    for _, source := range sources {
        for _, key := range []string{"options", "values", "items", "enum"} {
            raw, exists := source[key]
            if !exists {
                continue
            }

            candidates := make([]interface{}, 0)
            switch typed := raw.(type) {
            case []interface{}:
                candidates = append(candidates, typed...)
            case map[string]interface{}:
                if store, ok := typed["store"].([]interface{}); ok {
                    candidates = append(candidates, store...)
                }
            }

            if len(candidates) == 0 {
                continue
            }

            var preferred interface{}
            for _, item := range candidates {
                if option, ok := item.(map[string]interface{}); ok {
                    if def, ok := option["default"].(bool); ok && def {
                        preferred = item
                        break
                    }
                    if sel, ok := option["selected"].(bool); ok && sel {
                        preferred = item
                        break
                    }
                    if preferred == nil {
                        preferred = item
                    }
                } else if preferred == nil {
                    preferred = item
                }
            }

            selected := candidates
            if preferred != nil {
                selected = []interface{}{preferred}
            }

            for _, item := range selected {
                if option, ok := item.(map[string]interface{}); ok {
                    for _, candidate := range []string{"value", "id", "code"} {
                        if val, ok := option[candidate]; ok {
                            return val, true
                        }
                    }
                } else if item != nil {
                    return item, true
                }
            }
        }
    }
    return nil, false
}

File: samples/node/workflow-overview.js
'use strict';

/**
 *  MAKE CONNECTION WORKFLOW (ONE-PAGER)
 *
 *  This file is intentionally standalone and heavily commented so a developer can
 *  translate the logic to another language or platform without digging through
 *  the rest of the project. It mirrors the behaviour of the service in a
 *  streamlined, readable form.
 *
 *  The real implementation lives inside the src/ folder. This file keeps the
 *  same structure but swaps most implementation details for concise helpers and
 *  comments that describe the intent of each step.
 */

// ---------------------------------------------------------------------------
// 1. REQUIRED SETTINGS
// ---------------------------------------------------------------------------
const path = require('path');
const ROOT_DIR = path.resolve(__dirname, '..', '..');
require('dotenv').config({ path: path.resolve(ROOT_DIR, '.env') });
const fs = require('fs/promises');
const fetch = require('node-fetch');
const { spawn } = require('child_process');
const readline = require('readline');

const SETTINGS = {
  instanceUrl: process.env.INSTANCE_URL?.replace(/\/$/, ''),
  authToken: process.env.AUTH_TOKEN,
  teamId: process.env.TEAM_ID,
  blueprintDir: path.resolve(ROOT_DIR, 'blueprints'),
  updatedDir: __dirname,
  port: Number(process.env.PORT) || 777,
};

function ensureSettings() {
  const missing = Object.entries({
    INSTANCE_URL: SETTINGS.instanceUrl,
    AUTH_TOKEN: SETTINGS.authToken,
    TEAM_ID: SETTINGS.teamId,
  })
    .filter(([, value]) => !value)
    .map(([name]) => name);

  if (missing.length) {
    throw new Error(`Missing required env vars: ${missing.join(', ')}`);
  }
}

// ---------------------------------------------------------------------------
// 2. TOP-LEVEL WORKFLOW
// ---------------------------------------------------------------------------
/**
 *  runWorkflow
 *  ----------
 *  High-level orchestration that mirrors the production flow:
 *    - parse the blueprint and pick an app
 *    - collect required scopes for every relevant module
 *    - request the connection form and build the payload
 *    - create the connection + fetch consent URL
 *    - (optional) test the connection
 *    - save a remapped copy of the blueprint with the new connection id
 */
async function runWorkflow({ blueprintName, accountNameOverride }) {
  ensureSettings();

  const blueprint = await loadBlueprint(blueprintName);
  const { appName, moduleNames } = analyseBlueprint(blueprint);

  const appDefinition = await fetchJSON(`/api/v2/imt/apps/${appName}`);
  const scopes = collectScopes(appDefinition, moduleNames);

  const formSchema = await fetchJSON(`/api/v2/imt-forms/connections/create?type=${appName}&teamId=${SETTINGS.teamId}`);
  const payload = buildConnectionPayload({
    formSchema,
    appName,
    accountNameOverride,
    scopes,
  });

  const connection = await createConnection(payload);
  const consentUrl = await fetchConsentUrl(connection.id);

  const updatedBlueprintPath = await saveUpdatedBlueprint({
    originalName: blueprintName,
    blueprint,
    connectionId: connection.id,
  });

  return {
    connectionId: connection.id,
    consentUrl,
    payload,
    scopes,
    updatedBlueprintPath,
  };
}

// ---------------------------------------------------------------------------
// 3. BLUEPRINT HANDLING
// ---------------------------------------------------------------------------
async function loadBlueprint(name) {
  const candidates = [
    name,
    path.join(SETTINGS.blueprintDir, name || ''),
  ].filter(Boolean);

  for (const candidate of candidates) {
    try {
      const raw = await fs.readFile(candidate, 'utf8');
      return JSON.parse(raw);
    } catch (error) {
      if (error.code !== 'ENOENT') {
        throw error;
      }
    }
  }

  throw new Error(`Blueprint not found: ${name}`);
}

function analyseBlueprint(blueprint) {
  /**
   *  Blueprints store modules inside the `flow` array. Each module has a string
   *  like "hubspotcrm:getContact". We gather every module, skip the Facebook
   *  helper app, and target the first remaining app.
   */
  const allModules = [];

  function walk(node) {
    if (!node) return;
    if (Array.isArray(node)) {
      node.forEach(walk);
      return;
    }
    if (typeof node === 'object') {
      if (typeof node.module === 'string') {
        const [app, action] = node.module.split(':');
        if (app && action) {
          allModules.push({ app, action });
        }
      }
      Object.values(node).forEach(walk);
    }
  }

  walk(blueprint.flow);

  const filtered = allModules.filter(({ app }) => app !== 'facebook-conversion-leads');
  if (!filtered.length) {
    throw new Error('No eligible modules found in blueprint.');
  }

  const appName = filtered[0].app;
  const moduleNames = filtered.filter(({ app }) => app === appName).map(({ action }) => action);

  return { appName, moduleNames };
}

// ---------------------------------------------------------------------------
// 4. COLLECT SCOPES
// ---------------------------------------------------------------------------
function collectScopes(appDefinition, moduleNames) {
  /**
   *  The Make app definition groups modules into actions/searches/triggers.
   *  Each module has a `scopes` array that we merge together.
   */
  const buckets = ['actions', 'searches', 'triggers'];
  const lookup = new Map();

  buckets.forEach((bucket) => {
    const items = appDefinition.app?.[bucket] || [];
    items.forEach((item) => {
      if (item?.name) {
        lookup.set(item.name.toLowerCase(), item.scopes || item.scope || []);
      }
    });
  });

  const combined = new Set();
  moduleNames.forEach((name) => {
    const scopes = lookup.get(name.toLowerCase()) || [];
    scopes.forEach((scope) => combined.add(scope));
  });

  return [...combined];
}

// ---------------------------------------------------------------------------
// 5. BUILD CONNECTION PAYLOAD
// ---------------------------------------------------------------------------
function buildConnectionPayload({ formSchema, appName, accountNameOverride, scopes }) {
  /**
   *  The connection form schema describes all expected fields (accountName,
   *  accountType, property, custom scopes, etc.). We traverse the schema, fill
   *  defaults, apply overrides, and ensure required fields are not empty.
   */
  const fields = [];

  function collect(node) {
    if (!node) return;
    if (Array.isArray(node)) {
      node.forEach(collect);
      return;
    }
    if (typeof node === 'object') {
      const identifier = typeof node.key === 'string' && node.key.length
        ? node.key
        : typeof node.name === 'string' && node.name.length
          ? node.name
          : null;
      const resolvedType = node.type || node?.templateOptions?.type;
      if (identifier && !['button', 'content', 'htmlelement'].includes(String(resolvedType || '').toLowerCase())) {
        const resolvedDefault = node.defaultValue !== undefined
          ? node.defaultValue
          : node?.templateOptions?.defaultValue !== undefined
            ? node.templateOptions.defaultValue
            : node?.templateOptions?.default !== undefined
              ? node.templateOptions.default
              : node.default !== undefined
                ? node.default
                : undefined;
        fields.push({
          key: identifier,
          defaultValue: resolvedDefault,
          type: resolvedType,
          data: node.data,
          templateOptions: node.templateOptions,
        });
      }

      Object.entries(node).forEach(([propKey, value]) => {
        if (propKey === 'components') {
          return;
        }
        if (propKey === 'options' && value && typeof value === 'object' && Array.isArray(value.store)) {
          const store = value.store;
          if (store.length) {
            const selected = store.find((option) => option && (option.default === true || option.selected === true)) || store[0];
            if (selected) {
              collect(selected);
            }
          }
          return;
        }
        if (propKey === 'nested' && Array.isArray(value)) {
          value.forEach((entry) => collect(entry));
          return;
        }
        collect(value);
      });
    }
  }

  collect(formSchema);

  const payload = {};
  fields.forEach((field) => {
    const { key, defaultValue } = field;
    payload[key] = defaultValue !== undefined ? defaultValue : getFieldFallback(field);
  });

  if (typeof payload.accountName === 'string' && /{{.*}}/.test(payload.accountName)) {
    payload.accountName = '';
  }

  payload.accountType = payload.accountType || appName;
  payload.accountName =
    accountNameOverride ||
    payload.accountName ||
    `${appName.replace(/[-_]/g, ' ')} connection`;

  payload.customScopes = scopes;
  if (!payload.customScopes.length) {
    console.warn('customScopes is empty; the connection may lack permissions.');
  }

  if ('property' in payload && !payload.property) {
    throw new Error('The Make API expects the property field for this connection.');
  }

  return payload;
}

function getFieldFallback(field) {
  const type = typeof field.type === 'string' ? field.type.toLowerCase() : '';
  const templateType = typeof field?.templateOptions?.type === 'string'
    ? field.templateOptions.type.toLowerCase()
    : '';
  if (type === 'boolean' || templateType === 'boolean') {
    return false;
  }

  const dataValue = pickOptionFallback(field);
  if (dataValue !== undefined) {
    return dataValue;
  }

  return '';
}

function pickOptionFallback(field) {
  const sources = [field?.data, field?.templateOptions].filter(Boolean);
  if (!sources.length) {
    return undefined;
  }

  for (const source of sources) {
    if (typeof source === 'object' && Object.prototype.hasOwnProperty.call(source, 'defaultValue')) {
      return source.defaultValue;
    }
  }

  const optionLists = sources
    .flatMap((source) => (typeof source === 'object' ? extractOptionLists(source) : []));
  if (!optionLists.length) {
    return undefined;
  }

  const flattened = optionLists.flat();
  const preferred = flattened.find((option) => option && typeof option === 'object' && option.default === true)
    || flattened.find((option) => option && typeof option === 'object' && option.selected === true);

  const firstOption = preferred || flattened.find((option) => {
    if (option === null || option === undefined) {
      return false;
    }
    if (typeof option === 'object') {
      return option.value !== undefined || option.id !== undefined || option.code !== undefined;
    }
    return true;
  });

  if (!firstOption) {
    return undefined;
  }

  if (typeof firstOption === 'object') {
    return firstOption.value ?? firstOption.id ?? firstOption.code;
  }
  return firstOption;
}

function extractOptionLists(source) {
  if (!source || typeof source !== 'object') {
    return [];
  }

  const lists = [];
  ['options', 'values', 'items', 'enum'].forEach((key) => {
    const value = source[key];
    if (Array.isArray(value) && value.length) {
      lists.push(value);
    } else if (value && typeof value === 'object' && Array.isArray(value.store) && value.store.length) {
      lists.push(value.store);
    }
  });
  return lists;
}

// ---------------------------------------------------------------------------
// 6. MAKE API CALLS
// ---------------------------------------------------------------------------
async function createConnection(payload) {
  const body = JSON.stringify(payload);
  const response = await fetchJSON(`/api/v2/connections?teamId=${SETTINGS.teamId}&inspector=0`, {
    method: 'POST',
    body,
  });
  const connection = response?.connection || response;
  if (!connection || !connection.id) {
    throw new Error(`Connection response missing id: ${JSON.stringify(response)}`);
  }
  return connection;
}

async function fetchConsentUrl(connectionId) {
  const response = await fetchRaw(`/api/v2/oauth/auth/${connectionId}`, {
    method: 'GET',
    redirect: 'manual',
  });
  const location = response.headers.get('location');
  if (!location) {
    throw new Error('Consent URL missing in response headers.');
  }
  return location;
}

async function testConnection(connectionId) {
  return fetchJSON(`/api/v2/connections/${connectionId}/test`, {
    method: 'POST',
  });
}

// ---------------------------------------------------------------------------
// 7. UPDATED BLUEPRINT SAVING
// ---------------------------------------------------------------------------
async function saveUpdatedBlueprint({ originalName, blueprint, connectionId }) {
  const clone = JSON.parse(JSON.stringify(blueprint));
  let updated = 0;

  function rewrite(node) {
    if (!node) return;
    if (Array.isArray(node)) {
      node.forEach(rewrite);
      return;
    }
    if (typeof node === 'object') {
      if (
        typeof node.module === 'string' &&
        !node.module.startsWith('facebook-conversion-leads')
      ) {
        if (node.parameters && node.parameters.__IMTCONN__) {
          node.parameters.__IMTCONN__ = connectionId;
          updated += 1;
        }
      }
      Object.keys(node).forEach((key) => rewrite(node[key]));
    }
  }

  rewrite(clone.flow);

  await fs.mkdir(SETTINGS.updatedDir, { recursive: true });
  const baseName = path.basename(originalName, path.extname(originalName) || '.json');
  const targetPath = path.join(
    SETTINGS.updatedDir,
    `${baseName}-updated.json`
  );
  await fs.writeFile(targetPath, `${JSON.stringify(clone, null, 4)}\n`);

  console.log(`Updated ${updated} module(s) in ${targetPath}`);
  return targetPath;
}

// ---------------------------------------------------------------------------
// 8. LIGHTWEIGHT FETCH HELPERS
// ---------------------------------------------------------------------------
async function fetchJSON(relativeUrl, init = {}) {
  const response = await fetchRaw(relativeUrl, init);
  const text = await response.text();
  try {
    return JSON.parse(text);
  } catch (error) {
    throw new Error(`Expected JSON from ${relativeUrl} but received: ${text}`);
  }
}

function fetchRaw(relativeUrl, init = {}) {
  const url = `${SETTINGS.instanceUrl}${relativeUrl}`;
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `Token ${SETTINGS.authToken}`,
    ...(init.headers || {}),
  };
  return fetch(url, { ...init, headers });
}

// ---------------------------------------------------------------------------
// 9. SAMPLE USAGE
// ---------------------------------------------------------------------------
/**
 *  This block is optional. It demonstrates how to invoke the workflow directly
 *  from the command line:
 *
 *    node docs/workflow-overview.js "Hubspot - FBcapi.json" "My Friendly Name"
 */
function launchBrowser(url) {
  const platform = process.platform;
  const escape = (value) => value.replace(/"/g, '\\"');

  const attempts =
    platform === 'darwin'
      ? [
          { command: '/usr/bin/open', args: [url] },
          { command: '/usr/bin/open', args: ['-g', url] },
          { command: '/usr/bin/osascript', args: ['-e', `open location "${escape(url)}"`] },
        ]
      : platform === 'win32'
        ? [
            { command: 'cmd', args: ['/c', 'start', '', url] },
            { command: 'powershell', args: ['-NoProfile', 'Start-Process', url] },
          ]
        : [
            { command: 'xdg-open', args: [url] },
            { command: 'gio', args: ['open', url] },
            { command: 'gnome-open', args: [url] },
          ];

  function tryNext(list) {
    if (!list.length) {
      console.warn('Unable to launch browser automatically. Open the consent URL manually.');
      return;
    }

    const { command, args } = list[0];
    try {
      const child = spawn(command, args, { stdio: 'ignore', detached: true });
      child.on('error', () => tryNext(list.slice(1)));
      child.on('exit', (code) => {
        if (code === 0 || code === null) {
          child.unref();
        } else {
          tryNext(list.slice(1));
        }
      });
      child.unref();
    } catch (error) {
      tryNext(list.slice(1));
    }
  }

  tryNext(attempts);
}

function waitForEnter(message) {
  if (!process.stdin.isTTY) {
    return Promise.resolve();
  }
  return new Promise((resolve) => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question(message, () => {
      rl.close();
      resolve();
    });
  });
}

if (require.main === module) {
  const [blueprintName, accountNameOverride] = process.argv.slice(2);

  if (!blueprintName) {
    console.error('Usage: node workflow-overview.js <blueprint.json> [accountName]');
    process.exit(1);
  }

  runWorkflow({ blueprintName, accountNameOverride })
    .then(async (result) => {
      console.log('Connection created:', result.connectionId);
      console.log('Consent URL:', result.consentUrl);
      console.log('Updated blueprint:', result.updatedBlueprintPath);
      launchBrowser(result.consentUrl);
      console.log('If the consent page did not open automatically, open the URL above in your browser.');
      await waitForEnter('Complete the consent flow, then press Enter to trigger the Make test...');
      const testResponse = await testConnection(result.connectionId);
      console.log('Test response:', JSON.stringify(testResponse, null, 2));
    })
    .catch((error) => {
      console.error('Workflow failed:', error);
      process.exit(1);
    });
}

module.exports = {
  runWorkflow,
  testConnection,
};
